File: LinkedList/SingleLinkedList.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package linkedlist;

import java.util.NoSuchElementException;

/**
 *
 * @author Kakatuahitam
 */
public class SingleLinkedList<Type> {
    private SingleNode<Type> front;
    private int size;
    
    public SingleLinkedList(){
        front = null;
    }
    
    public void addFront(Type x){
        if (isEmpty()){
            front = new SingleNode<>(x);
        }
        
        else{
           SingleNode<Type> temp = front;
           front = new SingleNode<>(x, temp);
        }
        size++;
    }
    
    public void addEnd(Type x){
        if (isEmpty()){
            front = new SingleNode<>(x);
        }
        else{
            SingleNode<Type> temp = front;
            while(temp.next != null){
                temp = temp.next;
            }
            temp.next = new SingleNode<>(x, null);
        }
        size++;
    }
    
    public void addBefore(Type x, Type y){
        if(isEmpty()){
            throw new NoSuchElementException("List is empty");
        }
        
        SingleNode<Type> current = front;
        
        if (current.data == x){
            addFront(y);
        }
        else{
            SingleNode<Type> parent = null;
            
            while(current != null && !current.data.equals(x)){
                parent = current;
                current = current.next;
            }
            
            if (current != null){
                parent.next = new SingleNode(y, current);
                size++;
            }
            else{
                throw new NoSuchElementException("Node" + x.toString() + "not found");
            }
        }
    }    
            
    public void addAfter(Type x, Type y){
        if(isEmpty()){
            throw new NoSuchElementException("Element " + x.toString() + " not found");
        }
        
        SingleNode<Type> current = front;
        
        while(current != null && !current.data.equals(x)){
            current = current.next;
        }
        
        if(current == null){
            throw new NoSuchElementException("Element " + x.toString() + " not found");
        }
        
        SingleNode<Type> newNode = new SingleNode<>(y, current.next); 
        
        current.next = newNode;
        size++;
        
    }
    
    public void remove(Type x){
        if(isEmpty()){
            throw new NoSuchElementException("Element " + x.toString() + " not found");
        }
        
        if (front.data.equals(x)){
            front = front.next;
            return;
        }
        
        SingleNode<Type> current = front;
        SingleNode<Type> currbefore = new SingleNode<>(null, current);
        
        while(current != null && !current.data.equals(x)){
            current = current.next;
            currbefore = currbefore.next;
        }
        
        if(current == null){
            throw new NoSuchElementException("Element " + x.toString() + " not found");
        }
        
        currbefore.next = current.next;
        
        size--;
    }
    
    public boolean isEmpty(){
        return size == 0;
    }
    
    public int size(){
        return size;
    }
    
    @Override
    public String toString() {
        SingleNode<Type> temp = front;
        StringBuilder builder = new StringBuilder("[");
        
        while(temp != null){
            builder.append(temp.data).append(",");
            temp = temp.next;
        }
        
        return builder.toString() + "\b]";
    }
}

File: LinkedList/SingleLinkedList.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun  2 04:21:29 2020

@author: Kakatuahitam
"""
from SingleNode import SingleNode as sn

class SingleLinkedList():
    def __init__(self):
        self.front = None
        self.size = 0
        
    def __repr__(self):
        nodes = []
        curr = self.front
        while curr:
            nodes.append(repr(curr))
            curr = curr.next
        return '[' + ', '.join(nodes) + ']'
    
    def addFront(self, x):
        if self.isEmpty():
            self.front = sn(x)
        else:
            temp = self.front
            self.front = sn(x, temp)
        self.size += 1
        
    def addEnd(self, x):
        if self.isEmpty():
            self.front = sn(x)
        else:
            temp = self.front
            while temp.next != None:
                temp = temp.next
                
            temp.next = sn(x, None)
            
        self.size += 1
        
    def addBefore(self, x, y):
        if self.isEmpty():
            print("can't add", y, "before", x, "\nlist ls empty")
            
        current = self.front
        if current.data == x:
            self.addFront(y)
            
        else:
            parent = None
            
            while current != None and not(current.data == x):
                parent = current
                current = current.next
            
            if current != None:
                parent.next = sn(y, current)
                self.size += 1
            else:
                print("no node")
                return
    
    def addAfter(self, x, y):
        if self.isEmpty():
            print("can't add", y, "after", x, "\nlist ls empty")
        
        current = self.front
        while current != None and not (current.data == x):
            current = current.next
        
        if current == None:
            print("no", x,"node")
            
        newNode = sn(y, current.next)
        current.next = newNode
        self.size += 1
        
    def remove(self, x):
        if self.isEmpty():
            print("can't remove", x, "\nlist ls empty")
        
        if self.front.data == x:
            self.front = self.front.next
            return
        
        current = self.front
        currbefore = sn(next = current)
        
        while current != None and not (current.data == x):
            current = current.next
            currbefore = currbefore.next
            
        if current == None:
            print("no node")
            return
        
        currbefore.next = current.next
        self.size -= 1
      
    def isEmpty(self):
        return self.size == 0
    
    def wSize(self):
        return self.size